src/KeyRegistry.sol

```diff
@@ -0,0 +1,325 @@
+        signer.merkleRoot = merkleRoot;
+        emit Freeze(fid, scope, key, key, merkleRoot);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                        INITIAL MIGRATION
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Set the time of the signer migration and emit an event. Hubs will watch this event and
+     *         cut over to use the onchain registry as their source of truth after this timestamp.
+     *         Only callable by the contract owner.
+     */
+    function migrateSigners() external onlyOwner {
+        if (isMigrated()) revert AlreadyMigrated();
+        signersMigratedAt = uint40(block.timestamp);
+        emit SignersMigrated();
+    }
+
+    /**
+     * @notice Register multiple signers as part of the initial migration. Only callable by the contract owner.
+     *
+     * @param fids  A list of fids to associate with keys.
+     * @param keys  A list of public keys to register for each fid, in the same order as the fids array.
+     */
+    function bulkAddSignersForMigration(uint256[] calldata fids, bytes[][] calldata keys) external onlyOwner {
+        if (isMigrated() && block.timestamp > signersMigratedAt + gracePeriod) revert Unauthorized();
+        if (fids.length != keys.length) revert InvalidBatchInput();
+
+        unchecked {
+            for (uint256 i = 0; i < fids.length; i++) {
+                uint256 fid = fids[i];
+                for (uint256 j = 0; j < keys[i].length; j++) {
+                    _register(fid, 1, keys[i][j]);
+                }
+            }
+        }
+    }
+
+    /**
+     * @notice Remove multiple signers as part of the initial migration. Only callable by the contract owner.
+     *         Removal is not the same as revocation: this function sets the signer state back to UNINITIALIZED,
+     *         rather than REVOKED. This allows the owner to correct any errors in the initial migration until
+     *         the grace period expires.
+     *
+     * @param fids  A list of fids to whose registered keys should be removed.
+     * @param keys  A list of public keys to remove for each fid, in the same order as the fids array.
+     */
+    function bulkRemoveSignersForMigration(uint256[] calldata fids, bytes[][] calldata keys) external onlyOwner {
+        if (isMigrated() && block.timestamp > signersMigratedAt + uint40(gracePeriod)) revert Unauthorized();
+        if (fids.length != keys.length) revert InvalidBatchInput();
+
+        unchecked {
+            for (uint256 i = 0; i < fids.length; i++) {
+                uint256 fid = fids[i];
+                for (uint256 j = 0; j < keys[i].length; j++) {
+                    _remove(fid, 1, keys[i][j]);
+                }
+            }
+        }
+    }
+
+    function _register(uint256 fid, uint256 scope, bytes calldata key) internal {
+        Signer storage signer = signers[fid][scope][key];
```

Comment:
Hmm, so same signer can exist with multiple scopes? This is going to make the hub logic more complicated I think we should keep it to a single scope per key.

Reply:
going to fix this